feat(all): add request id generation

jjaracz opened the pull request:
## What I did?

I added:
- request id to logger
- creation of request id on nestjs interceptor
- new decorator called `OgmaLoggerRequestScoped`
- options to `forFeature` in `OgmaModule` to be able to register request scoped provieder
- on abstract interceptor (as on the classes that it extends) `setRequestId`

## Why I did it?

I liked your library but I need to be able to automatically add request id to each created log, so I decide to implement it on my way and make pull request in hope you would like this solution.

## What needs to be done?

- [ ] update README.md with information about how to use request scoped logger with **bolded** or something that when you want to use request scoped logger then you need to have in mind that each service that use that service and service itself will be request scoped, so for each new request which will come, new instance of each service will be created.
- [x] add tests for that feature
- [x] fix tests

jjsinch committed:
feat(all): add request id generation and log

jmcdo29 commented on packages/nestjs-module/package.json:
```diff
@@ -38,6 +38,7 @@
   "peerDependencies": {
     "@nestjs/common": "^7.0.0",
     "@nestjs/core": "^7.0.0",
+    "@nestjs/microservices": "^7.0.0",
```
The module package itself shouldn't have a dependency on any of the `@nestjs/` packages other than `common` and `core`.

jmcdo29 replied:
I see that the reason for the dependency is the context that comes from the `@nestjs/microservices` package. Is there anything we can do to possibly make this not required? I'd rather not force users to install `@nestjs/microservices` to use the package, or have to enable `skipLibCheck` as I know that many people prefer to have that when it comes to strict mode.

jjaracz replied:
Yeah, it was only for context and type for it. To remove it from there we could easly just create our interface for that and use request instead, because Nest inject request object in there if you use that in case of http but when you use that in case some other protocol like grpc then you get RequestContextHost, so we could create our types for that as it could be request or context and then I think we could remove that dependency from there.

jmcdo29 commented on packages/nestjs-module/src/interceptor/ogma.interceptor.ts:
```diff
@@ -86,10 +90,21 @@ export class OgmaInterceptor implements NestInterceptor {
     );
   }
+
+  public generateRequestId(): string {
```
Was this made as a class method so that it could be overridden in other implementations? If so, the `OgmaInterceptor` should be removed from the `OgmaModule` when it comes to being bound via `APP_INTERCEPTOR` otherwise two interceptors for logging will be created. This will mean the user then needs to bind the `OgmaInterceptor` on their own, which would be a breaking change from the current functionality (though one I would be okay with).

 The other option is we allow a function to be passed as an option instead, then call `this.options.generateRequestId()`, but I like this class based approach.

jmcdo29 commented on packages/nestjs-module/src/ogma.provider.ts:
```diff
@@ -3,12 +3,14 @@ import {
+  Scope,
 } from '@nestjs/common';
-import { Reflector } from '@nestjs/core';
+import { ModuleRef, Reflector } from '@nestjs/core';
```
Can we remove `ModuleRef` if it isn't used?

jmcdo29 commented on packages/nestjs-module/src/ogma.provider.ts:
```diff
@@ -105,6 +113,26 @@ export function createLoggerProviders(
+  const token = createRequestScopedProviderToken(topic);
+  return [
+    {
+      inject: [OGMA_INSTANCE, CONTEXT],
```
Why use `CONTEXT` instead of `REQUEST`?

jjaracz replied:
Actually there is no difference as `CONTEXT` as you could see in sources it is in fact string `REQUEST`. But I used `CONTEXT` because `REQUEST` sounds for me like it is for http based queries and I didn't want any misunderstandings as that should handle requests for all protocols not just for http (for example on grpc we have `metadata` and `payload` and we get `metadata` from `ExecutionContext` by method `getContext`). But when we decide to remove `microservice` dependency then we will be forced to use `REQUEST` instead of `CONTEXT`

jmcdo29 replied:
What we can do, if you still want to use the word `CONTEXT` to help keep things clearer, is make our own `const CONTEXT = REQUEST` and use that constant. They're all the same anyways. Or just `import { REQUEST as CONTEXT }`

jmcdo29 commented on packages/nestjs-module/src/ogma.service.ts:
```diff
@@ -23,91 +25,118 @@ export class OgmaService implements LoggerService {
     this.ogma = ogma ?? new Ogma();
   }
 
+  private getRequestId(
```
Does this handle the cases for GQL or WS?

jjaracz replied:
Honestly I checked only GRPC and HTTP but GRPC should works on Nest just like GQL and WS. But yeah, that is thing to test before it will be merged and published

jmcdo29 commented on tsconfig.json:
```diff
@@ -12,7 +12,8 @@
     "lib": ["ES7"],
-    "esModuleInterop": true
+    "esModuleInterop": true,
+    "skipLibCheck": true
```
What's the reason for adding in the `skipLibCheck`?

jjaracz replied:
I pushed that by mistake as something didn't work for me as I don't quite understand how all of libs are connected and linked to each other. I will remove that in next commit with fix tests

jmcdo29 reviewed (changes requested):
Overall, looks like a really solid addition. I've got a couple of questions and requests.

1) If there's an error in a middleware or guard, the interceptor doesn't handle that. This is already a known feature, but how do we handle the request ID in that case?

2) Can you remove the `package-lock.json` files? The only lock file should be the `yarn.lock` at the root of the project.

3) Tests will obviously need to be updated for this. Is that something you can do as well, and you're just waiting to see my response to the PR before going through the effort to fix them, or do you need some help with them?

jjsinch committed:
test(module): fix tests regarding request id changes

jjsinch committed:
test: fix integration tests

jjsinch committed:
feat(module): remove registration of global interceptor

jjaracz commented:
Hi, @jmcdo29 i fixed almost all tests but integration http test fail but on my local configuration it passes. Do you know what could causing this issue? Also I removed global interceptor as you mention before.

I have an idea to remove that ugly `getInterceptor` method. We could override `OgmaInterceptor` with overwritten `log` method and inject it to test module.

jjsinch committed:
test: add tests for request id

jmcdo29 commented on packages/nestjs-module/src/ogma-core.module.ts:
```diff
@@ -42,16 +41,6 @@ export class OgmaCoreModule extends createConfigurableDynamicRootModule<
       useFactory: createOgmaServiceOptions,
       inject: [OGMA_OPTIONS],
     },
-    {
-      provide: APP_INTERCEPTOR,
-      useFactory: createOgmaInterceptorFactory,
-      inject: [
-        OGMA_INTERCEPTOR_OPTIONS,
-        OgmaService,
-        DelegatorService,
-        Reflector,
-      ],
-    },
```
I assume you've removed this to put the necessity on the dev to attach the OgmaInterceptor?

jjaracz replied:
Yes, I assume that everyone who would like to use `OgmaInterceptor` need to inject it by themself by eg.

```ts
@Module({
  ...
  providers: [
    {
      provide: APP_INTERCEPTOR,
      useClass: OgmaInterceptor
    }
  ]
})
export class MyModule {}
```

or by overwriting `OgmaInceptor` and inject self created one

jmcdo29 replied:
This is a great addition. As nice as it is to have the interceptor applied automatically, having it be applied by the user is definitely going to clean some stuff up. We will need to make a note of this in the readme of course. and I believe you didn't change the module options, right? So if no adapter is applied and `false` is not set on the `interceptor` option there's still a problem, right?

jjaracz replied:
Yeah I didn't change that. I was trying to move option `interceptor` into constructor of `OgmaInterceptor` but that change was huuuuuge and turns out that need to be changed delegate service and few others concepts. I didn't want to change something that you wouldn't like as that change impact on almost whole package.

There is no option to not set interceptor in options as it's required by types. But yeah I didn't touch options on `forRoot` and `forAsyncRoot`

jmcdo29 replied:
Perfect. I think overall, I'd like to keep that part of the functionality the same, just because if an adapter wasn't registered to begin with, and the interceptor was bound, there could be all sorts of other errors. Still can be, but now it's less likely.

jmcdo29 commented on packages/nestjs-module/src/ogma.service.ts:
```diff
@@ -23,24 +25,36 @@ export class OgmaService implements LoggerService {
+   * @param requestId Optional id of an request
    */
   public error(message: any, context?: string): void {
     this.printMessage(message, 'error', context);
```
There's no `requestId` on the `error` method. The docs decorator is there though, is this a miss?

jjaracz replied:
Yeah I forgot to mention. I found out that Nest.JS on error by default on third parameter pass name of error so in our scenario sets `requestId` which was unwanted so I decide to remove `requestId` from error cause we already have `printError` so if someone would like to print error with custom `requestId` could use that method instead. Do you have better idea to solve that issue? I couldn't find better

jmcdo29 replied:
I think this is fine. Having the `requestId` in `printError` works well for me

jjaracz commented:
I added test for that feature. The only thing that left is fix integration test (locally it works but on CI unfortunately not) and update `README.md`. Could I ask you to do that? I'm not a good writer and after I would add there that probably you would change it anyway.

- We need update it with breaking change of `OgmaInterceptor` (also I tagged `feat` commit with breaking change in git history) to inform people to register global interceptor by themself, they also could overwrite `OgmaInterceptor` `generateRequestId` method.
- We need to introduce new decorator `OgmaLoggerRequestScoped` and options in `OgmaModule.forFeature` with flag `addRequestId` and bold that when they would like to use that functionality need have in mind each service that would use that will be request scoped and also each service that use that service.

Is there something I forgot to fix or change?

jmcdo29 commented on packages/nestjs-module/src/ogma.service.ts:
```diff
@@ -50,64 +64,77 @@ export class OgmaService implements LoggerService {
+    if (this.requestContext && this.context) {
+      requestId = this.getRequestId(this.requestContext);
+    }
+    this.ogma[levelString](message, context, undefined, requestId);
```
What happens if a requestId is passed but the `OgmaService` also has a `this.context` and `this.requestContext`? Would that ever be a valid use case?

jjaracz replied:
Yeah, I will add there third condition with checking if `requestId` is not passed. Right now it will replace it in that case

jmcdo29 reviewed (changes requested):
Okay, other than the few comments this is looking pretty good! I think I know where the error is in the CI/CD pipeline, I'll work on getting that fixed today. I'm pretty sure it comes from `@nestjs/platofrn-fastify@7.4.2` where `fastify v3` became supported. Fastify v^2 is required in the devDeps, so Ithink I need to update that and we should be good 😸 thanks again for all the help!

jjsinch committed:
fix(module): add requestId check to printMessage

jmcdo29 commented:
All right! This is mostly ready to go! I'll work on updating the README, and getting the build fixed. There's a problem with axios at the moment where version 0.2.0 is expecting to have `@types/DOM` installed and that doesn't work for Node projects. And while Ogma can be used on either front or backend I made it with Node in mind, so that needs to be taken into account. I'll work on fixing the build later today/tomorrow when I have the time. Hoping to have this merged in soon. Thanks again, I'm looking forward to the addition

jmcdo29 commented:
Oh, and really great job keeping the existing singleton scope available while giving an **optional** request scope. That's awesome in my opinion.

jmcdo29 commented:
Okay! Everything is looking good again in the build. If you could, rebase from master to get the latest fix (updating fastify to v3) and we should be good to merge. I'll update the docs soon after that and publish a new release by the end of the week

jjaracz commented:
Yupi, CI pass successfully 🎉 

Thank you for code review and waiting for release!

jmcdo29 commented:
```
[2020-09-05T17:20:11.792Z] [jay-Lenovo] [tabletop] 584512 1599326411787770 [AppController#sayHello] [INFO] | ::ffff:127.0.0.1 - GET /api HTTP/
```
Works great locally! Thanks again!
